docs/Developers/Design/Upgrade_design.md

```diff
@@ -0,0 +1,146 @@
+The operator is responsible for incorporating the domain-specific expertise
+necessary to perform upgrades of the Gluster cluster. An administrator, either
+directly or through an upgrade service, will choose the desired version of
+Gluster (GCS) that should be running on the cluster, and the operator must
+properly and autonomously upgrade all components to match that request.
+
+# Problem overview
+
+A version of GCS is comprised of a specific set of container images plus the
+associated object manifests required to deploy them. In order to upgrade, the
+container images must be changed to those in a subsequent release. These
+updates, however, must be mindful of version compatibility between the
+components. Particularly when upgrading across a series of releases, it may be
+necessary to take intermediate upgrade steps to ensure client (CSI driver)
+versions remain compatible with server (Gluster container) versions.
```

Comment:
The client update has some limitation. The disturbance of fuse mounts could be an issue.

Reply:
Interesting point; thanks for bringing that up.
I suspect we'll see `transport endpoint not connected` when the CSI pod gets restarted on a node. I wonder if this means we need to drain nodes before upgrading the clients (and the implications for doing that in a rolling fashion w/ a DaemonSet).